### ISSUE TITLE
multiple-optimizations-after-first-at-scale-training-run

### ISSUE THREAD

**Issue by david-thrower** [david-thrower/HelixLM#2]
**GitHub Issue: Training Pipeline Instability and NaN Losses in 2-Stage Training (`tiny` preset, `seq_len=256`)**

## Summary
Running the 2-stage training pipeline (`train_helixlm_2stage.py, separate repo for now`) on the `tiny` model preset with `seq_len=256` and the full GPT-2 tokenizer (vocab 50,257) resulted critical, reproducible failures: **(1)** NaN losses emerging around batch 70, **(2)** catastrophically unstable perplexity (PPL "islands" of ~13 amidst spikes of 600+) caused by dataset chunking artifacts, and **(3)** inability to scale effective batch size on modest hardware due to missing gradient accumulation. The existing `quick_demo_cpu.py` (seq_len=96, no AMP, pre-truncated stories) trains perfectly, confirming the model architecture is sound and the failures are strictly in the training pipeline.

---

## Environment
- **Model:** `tiny` (actual params: ~19.8M with GPT-2 vocab; documentation table assumes ~100-token vocab)
- **Sequence length:** 256
- **Batch size:** 4
- **Dataset:** `david-thrower/HelixLM-tiny-5.0Mt-9125pt-715it-20260427`
- **Device:** CUDA with `torch.cuda.amp` enabled (default path in existing trainer, T4 Google Colab GPU instance)
- **Script:** `train_helixlm_2stage.py`

---

### Symptom
Training proceeds normally until batch 70, then loss abruptly becomes `NaN` and never recovers:
```
Batch 60/15687 | Loss: 9.7632 | PPL: 17381.35 | LR: 3.05e-04
Batch 70/15687 | Loss: nan | PPL: nan | LR: 3.55e-04
```

### Root Cause
Under `torch.cuda.amp`, `LinearAttnNode.forward()` performs cumulative sums over 256 positions with feature maps that grow to ~1.5–2.0 magnitude after warmup batches. The einsum `bhTf,bhTfd->bhTfd` and subsequent cumulative sum over `T=256` exceeds the `float16` max value (65,504), producing `inf` in the forward pass, which corrupts the backward pass into `NaN`.

### Fix
- **`helix_lm/nodes.py`:** Force the linear attention feature maps, einsum, and cumulative sums into `float32` unconditionally, then cast back to the input dtype at the end of `LinearAttnNode.forward()`.
- **`helix_lm/trainer.py`:** Make AMP **opt-in** (`use_amp=False` by default) and upgrade to the modern `torch.amp` API (`torch.amp.GradScaler("cuda")`, `torch.amp.autocast(...)`).

---

### Symptom
Loss and PPL behave like a "pinball machine" with extreme variance—e.g., PPL dropping to 24–75 on some batches while spiking to 400–600+ on adjacent batches:
```
Batch 370/417 | Loss: 2.6229 | PPL: 13.78
Batch 380/417 | Loss: 4.8840 | PPL: 132.15
Batch 400/417 | Loss: 3.2045 | PPL: 24.64
```

### Root Cause
The existing `HelixDataset` concatenates all documents into one contiguous token stream, then slices with `stride=128` and `seq_len=256`. This causes two distinct pathologies:
1. **50% training waste:** Every non-first chunk masks its first 128 labels as `-100`, discarding half the training signal.
2. **Document-boundary crossings:** Chunks frequently splice unrelated documents mid-sequence (e.g., math → poetry → code). The model is asked to predict the start of a new document from the tail of a completely unrelated one, producing impossible loss spikes. The "islands" of low PPL are interior chunks fully inside coherent documents.

### Fix
- **`helix_lm/dataset.py`:** Introduce `DocumentAwareDataset` and `create_document_loader()`.
  - Chunks **per-document only**, never crossing boundaries.
  - Uses **non-overlapping** chunks (`stride = seq_len`), eliminating label masking waste.
  - Only masks **padding** tokens in labels (`-100`).
  - Drops short documents/tails below `min_tail_len`.
  - Marks the final chunk of each document as `is_natural_stop=True` for future ACT integration.
- **`train_helixlm_2stage.py`:** Uses stage-specific `min_tail_len`:
  - Stage 1 (pretrain): `seq_len // 4` (drop noisy short docs/tails).
  - Stage 2 (instruct): `1` (preserve all conversation samples, even short ones).

---

### Symptom
Effective batch size is limited to the physical GPU batch size (4), making training noisy and preventing stable convergence at higher learning rates.

### Fix
- **`helix_lm/trainer.py`:** Complete rewrite adding `grad_accum_steps` support.
  - Loss is scaled by `1/grad_accum_steps` before `backward()`.
  - Optimizer and scheduler steps count **optimizer steps**, not raw batches.
  - Enables effective batch sizes of 32+ on consumer GPUs.

---

### Symptoms
- `FutureWarning` spam from deprecated `torch.cuda.amp.GradScaler` and `torch.cuda.amp.autocast`.
- AdamW `betas=(0.9, 0.95)` chased every noisy batch.
- `weight_decay=0.1` was too aggressive for a 5M-token dataset.
- `lr=5e-4` with batch=4 was too low for stable convergence.

### Fixes
- **Trainer:** Upgrade to `torch.amp` modern API.
- **Trainer:** Switch AdamW betas to standard `(0.9, 0.999)`.
- **`train_helixlm_2stage.py`:** Expose all non-structural hyperparameters as CLI overrides for Optuna sweeps:
  - `--grad-accum-steps`, `--warmup-steps`, `--weight-decay`, `--grad-clip`, `--eval-every`, `--use-amp`, `--min-tail-len`, etc.
- **Updated `TRAINING_CONFIGS` for `tiny`:**
  - `grad_accum_steps: 8` (effective batch = 32)
  - `lr_stage1: 1e-3`
  - `warmup_steps: 500`
  - `weight_decay: 0.01`
  - `epochs_stage1: 5`

---

## Files Changed

| File | Change |
|------|--------|
| `helix_lm/nodes.py` | `LinearAttnNode.forward()` computes attention math in `fp32`, casts back |
| `helix_lm/dataset.py` | New `DocumentAwareDataset` + `create_document_loader()` |
| `helix_lm/trainer.py` | Gradient accumulation, configurable AMP, NaN/Inf skipping, modern `torch.amp`, `DocumentAwareDataset` integration |
| `train_helixlm_2stage.py` | CLI-exposed hparams, stage-specific `min_tail_len`, updated `TRAINING_CONFIGS` |
| `helix_lm/test_long_sequences.py` | **New** smoke test verifying long-sequence stability, AMP safety, and grad accumulation |
| `helix_lm/__init__.py` | Export new dataset classes |

---

## Testing & Verification
- **`smoke_test.py`** (regression): Passes. Loss decreases 3.48 → 0.73, PPL < 5.0.
- **`test_long_sequences.py`** (new): Passes. Document-aware chunking verified (25 chunks, all exact length, only padding masked). Long-document training shows no NaN/Inf. Gradient accumulation runs cleanly. PPL remains stable and comparable to baseline.

---

## Backward Compatibility Notes
- **Breaking:** `Trainer.__init__()` signature changed (new keyword args: `grad_accum_steps`, `use_amp`, `min_tail_len`). Code using positional arguments beyond `generated_example_length` must switch to keyword args.
- **Behavioral:** Trainer now defaults to `DocumentAwareDataset`. This is intentional—training behavior changes from overlapping contiguous streams to per-document non-overlapping chunks.
- **Non-breaking:** Old `HelixDataset` and `create_helix_dataloader` remain in the codebase and continue to work.

---

## Recommended Training Command (post-fix)
```bash
python train_helixlm_2stage.py \
  --model-size tiny \
  --dataset-path <your-dataset> \
  --batch-size 4 \
  --grad-accum-steps 8 \
  --lr-stage1 1e-3 \
  --lr-stage2 3e-4 \
  --warmup-steps 500 \
  --weight-decay 0.01 \
  --epochs-stage1 5 \
  --epochs-stage2 2
  # Do NOT use --use-amp for tiny/small models (unnecessary and risky)
```

---

## Checklist
- [x] NaN root cause identified and fixed (fp32 cumulative sums in `LinearAttnNode`)
- [x] Dataset chunking fixed (no boundary crossings, no masked-label waste)
- [x] Gradient accumulation implemented
- [x] AMP made opt-in with modern `torch.amp` API
- [x] All hyperparameters exposed for Optuna tuning
- [x] Regression tests pass
- [x] New long-sequence smoke test passes
- [x] Update README parameter count table to reflect actual GPT-2 vocab sizes (future docs fix)
- [ ] Verify `HelixForCausalLM` weight tying behavior (noted but out of scope for this pipeline fix)
